src/main/java/org/jenkinsci/plugins/docker/workflow/WithContainerStep.java

```diff
@@ -349,9 +349,14 @@ private static class Decorator extends LauncherDecorator implements Serializable
                 }
                 @Override public void kill(Map<String,String> modelEnvVars) throws IOException, InterruptedException {
                     ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                    ByteArrayOutputStream baos2 = new ByteArrayOutputStream();
                     String executable = getExecutable();
-                    if (getInner().launch().cmds(executable, "exec", container, "ps", "-A", "-o", "pid,command", "e").stdout(baos).quiet(true).start().joinWithTimeout(DockerClient.CLIENT_TIMEOUT, TimeUnit.SECONDS, listener) != 0) {
-                        throw new IOException("failed to run ps");
+                    // TODO: Two issues cause this code to fail to kill anything when the `ps` binary in the container is not from procps.
+                    // 1: Other implementations of ps only include `jsc=<cookie here>` and `JENKINS_SERVER_COOKIE=$jsc` in their output.
+                    //    procps's ps also includes `JENKINS_SERVER_COOKIE=<cookie here>`, which is required for this code to work.
+                    // 2: `-o command` is not portable. `-o args` is more portable and seems to be equivalent, but `-o` itself is not necessarily portable, see JENKINS-52881.
+                    if (getInner().launch().cmds(executable, "exec", container, "ps", "-A", "-o", "pid,command", "e").stdout(baos).stderr(baos2).quiet(true).start().joinWithTimeout(DockerClient.CLIENT_TIMEOUT, TimeUnit.SECONDS, listener) != 0) {
+                        throw new IOException("failed to run ps. stderr: " + baos2.toString(Charset.defaultCharset().name()));
```

Comment:
This is needed to understand _why_ `ps` is failing. In my case I saw `-o command` not being supported by one implementation of `ps` as well as `ps` not being present in the container at all.